### PR TITLE
Refactor functions and unit test

### PR DESCRIPTION
**Pull request by juansep1231**
Added functions to get the sum of all repos' stars

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,11 +4,15 @@
...
   },
   "author": "",
   "license": "ISC",
   "dependencies": {
-    "axios": "^1.6.3"
+    "axios": "^1.6.3",
+    "jest": "^29.7.0"
+  },
+  "devDependencies": {
+    "jest-environment-jsdom": "^29.7.0"
```
Comment (dalejo96) [inline package.json]: Why this one?
Jest must be part of `devDependencies`

File: scripts/functions.js [hunk 2]
```diff
@@ -1,66 +1,69 @@
...
+};
 
-  return data.filter(repo => repo.stargazers_count > 5)
-}
-
-
-const getReposMoreThanFiveStars = () => { 
-
-  return getData().then(data => {
+const getReposMoreThanFiveStars = () => {
```
Comment (dalejo96) [inline scripts/functions.js]: Make the functions inside this file pure, please.

File: scripts/functionsAsyncAwait.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    return { name: repo.name, updated_date: repo.updated_at };
+  });
+  return lastFiveRepos;
+};
+
+const getTopFive = (dataArray) => {
+  return dataArray.slice(0, 5);
+};
+
+const sortDateData = (dataArray) => {
```
Comment (dalejo96) [inline scripts/functionsAsyncAwait.js]:
```suggestion
const sortDataByDate = (dataArray) => {
```
What do you think?

File: scripts/main.js [hunk 4]
```diff
@@ -1,6 +1,16 @@
...
\ No newline at end of file
+getArrayOfData()
+  .then((data) => {
+    console.log(getReposMoreThanFiveStars(data));
+    console.log(getLastUpdatedRepos(data));
+    console.log(getSumOfReposStars(data));
+  })
+  .catch((error) => {
+    console.log(error);
+  });
```
Comment (dalejo96) [inline scripts/main.js]: You can use currying here!
```suggestion
  .catch(console.log);
```

File: test/unit/functionsAsyncAwait.test.js [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  sortDateData,
+  getSum,
+  getSumOfReposStars,
+  getReposMoreThanFiveStars,
+  getLastUpdatedRepos,
+} = require("../../scripts/functionsAsyncAwait.js");
+
+const { data } = require("../fixtures/data.js");
+
+describe("Test of functions to get the repos with more than five stars", () => {
```
Comment (dalejo96) [inline test/unit/functionsAsyncAwait.test.js]: It is kind of common to have the first describe as the function name that you are testing in this case:
```suggestion
describe("getFilteredsStargazers", () => {
```
Do this for all the tests please

File: test/unit/functionsAsyncAwait.test.js [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+  test("returns an array with all information about the repos with more than 5 stars", () => {
```
Comment (dalejo96) [inline test/unit/functionsAsyncAwait.test.js]: WDYT?
```suggestion
  test("returns an array with repos with more than 5 stars", () => {
```
Test descriptions must be short and clear
Do this for all the tests

File: test/unit/functionsAsyncAwait.test.js [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+});
+
+describe("Test of functions to get the sum of all the repos stars", () => {
+  test("returns the sum of all the repos stars", () => {
+    expect(getSumOfReposStars(data)).toEqual(61);
+  });
+});
+
+/*
+jest.mock("./functions");
```
Comment (dalejo96) [inline test/unit/functionsAsyncAwait.test.js]: Delete unused code

**Review by dalejo96 — CHANGES_REQUESTED** [review]
This is looking great, please add [Cris](https://github.com/CristopherDavid) as repo contributor and assign him as a reviewer for this PR. 
Check the comments and let me know if you have any questions.

File: scripts/functions.js (hunk 2 above)
Comment (juansep1231) [reply scripts/functions.js]: Could I just delete this file, because I think that the functions would be the same as the ones in the functionsAsyncAwait.js file.

File: package.json [hunk 8]
```diff
@@ -4,11 +4,15 @@
...
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest"
   },
   "author": "",
   "license": "ISC",
   "dependencies": {
     "axios": "^1.6.3"
+  },
+  "devDependencies": {
+    "jest-environment-jsdom": "^29.7.0",
```
Comment (dalejo96) [inline package.json]: remove this dependency

File: scripts/functionsAsyncAwait.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  } catch (error) {
+    console.log(error);
+  }
+};
+
+// FUCNTIONS TO RETURN THE REPOS WIHT MORE THAN FIVE STARS
+
+const getArrayOfData = async () => {
+  const data = await getData();
+  return (dataArray = Object.values(data));
```
Comment (dalejo96) [inline scripts/functionsAsyncAwait.js]:
```suggestion
  return Object.values(data);
```

File: scripts/functionsAsyncAwait.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+const getSumOfReposStars = (dataArray) => {
+  const sumOfStars = getSum(dataArray);
+  return sumOfStars;
+};
+
+const getSum = (dataArray) => {
+  return dataArray.reduce((acc, repo) => acc + repo.stargazers_count, 0);
+};
+
```
Comment (dalejo96) [inline scripts/functionsAsyncAwait.js]: Place the reduce inside the `getSumOfReposStars`

File: test/unit/functionsAsyncAwait.test.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+      {
+        name: "repo10",
+        stars: 8,
+      },
+    ]);
+  });
+});
+
+describe("getLastUpdatedRepos", () => {
+  test("returns an array with the top five repos", () => {
```
Comment (dalejo96) [inline test/unit/functionsAsyncAwait.test.js]: Add another test when the data is an empty array.
```
test("returns an array with the top five repos when data array is filled with info", () => {
test("returns an empty array with data array is empty", () => {
test("returns an empty array with data array is null", () => {
test("returns an array with the repos when data array is filled with just 3 items", () => {
```
Do this for all the functions

**Review by dalejo96 — CHANGES_REQUESTED** [review]
Add more use cases to the tests and validate `isUndefined`

File: scripts/functionsAsyncAwait.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    return await response.json();
+  } catch (error) {
+    console.log(error);
+  }
+};
+
+//FUNCTION TO CHECK IF THE DATA IS EMPTY, NULL OR UNDEFINED
+const checkEmptyData = (dataArray) => {
+  return dataArray === undefined || dataArray === null || dataArray.length <= 0;
+};
```
Comment (CristopherDavid) [inline scripts/functionsAsyncAwait.js]: Use implicit return get rid of the `{}` and the `return
`

File: scripts/functionsAsyncAwait.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+//FUNCTION TO EXCEUTE A FUNCTION IF DATA IS NOT EMPTY, NULL OR UNDEFINED
+const checkEmptyDataAndExecuteFunction = (dataArray, executeFunction) => {
+  if (checkEmptyData(dataArray)) {
+    return [];
+  }
+  return executeFunction(dataArray);
+};
```
Comment (CristopherDavid) [inline scripts/functionsAsyncAwait.js]: Use ternary operator and implicit return

File: test/unit/functionsAsyncAwait.test.js [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+    expect(getFilteredsStargazers(data)).toEqual([
+      {
+        name: "repo1",
+        stargazers_count: 10,
+        updated_at: "2020-11-18T12:00:00Z",
+      },
+      {
+        name: "repo5",
+        stargazers_count: 9,
+        updated_at: "2024-11-18T12:00:00Z",
```
Comment (CristopherDavid) [inline test/unit/functionsAsyncAwait.test.js]: Add a description to specify the case when this happens
```suggestion
describe("getReposMoreThanFiveStars", () => {
  describe("when an array of valid data is passed",() => {
   test("returns an array with repos with more than 5 stars", () => {
  
```

**Review by CristopherDavid — CHANGES_REQUESTED** [review]
Left some comments, great work

File: scripts/functionsAsyncAwait.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    getFilteredsStargazers(data).map((repo) => ({
+      name: repo.name,
+      stars: repo.stargazers_count,
+    }))
+  );
+
+// FUCNTIONS TO RETURN THE 5 LAST UPDATED REPOS
+
+const getLastUpdatedRepos = (dataArray) =>
+  checkEmptyDataAndExecuteFunction(dataArray, (data) => {
```
Comment (dalejo96) [inline scripts/functionsAsyncAwait.js]:
```suggestion
  checkEmptyDataAndExecuteFunction(dataArray, (data) => {
  if (isUndefined(dataArray)) return []
  return something(dataArray)
  
```

**Review by dalejo96 — APPROVED** [review]
Just use a simple validation and this is ready to go!

**Review by CristopherDavid — APPROVED** [review]
